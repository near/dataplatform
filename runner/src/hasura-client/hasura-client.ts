import fetch, { type Response } from 'node-fetch';
import pluralize from 'pluralize';

interface Dependencies {
  fetch: typeof fetch
}

interface SqlOptions {
  readOnly: boolean
  source?: string
}

export type HasuraPermission = 'select' | 'insert' | 'update' | 'delete';

interface TableDefinition {
  name: string
  schema: string
}
export interface HasuraRolePermission {
  role: string
  permission: {
    check?: Record<string, any>
    columns?: string[]
    backend_only?: boolean
    filter?: Record<string, any>
    allow_aggregations?: boolean
  }
}
export interface HasuraTableMetadata {
  table: TableDefinition
  insert_permissions?: HasuraRolePermission[]
  select_permissions?: HasuraRolePermission[]
  update_permissions?: HasuraRolePermission[]
  delete_permissions?: HasuraRolePermission[]
}

export interface HasuraDatabaseConnectionParameters {
  password: string
  database: string
  username: string
  host: string
  port: number
}

interface HasuraDatabaseUrl {
  connection_parameters: HasuraDatabaseConnectionParameters
}

interface DefaultHasuraDatabaseUrl {
  from_env: string
}

function isDefaultDatabaseUrl (
  object: DefaultHasuraDatabaseUrl | HasuraDatabaseUrl
): object is DefaultHasuraDatabaseUrl {
  return 'from_env' in object;
}

export interface HasuraConfiguration {
  connection_info: {
    database_url: DefaultHasuraDatabaseUrl | HasuraDatabaseUrl
    isolation_level?: string
    pool_settings?: Record<string, any>
    use_prepared_statements?: boolean
  }
}

export interface HasuraSource {
  name: string
  kind: string
  tables: HasuraTableMetadata[]
  configuration: HasuraConfiguration
}

export interface HasuraMetadata {
  version: number
  sources: HasuraSource[]
}

type MetadataRequestArgs = Record<string, any>;

type MetadataRequests = Record<string, any>;

interface Config {
  adminSecret: string
  endpoint: string
  pgHost: string
  pgHostHasura?: string
  pgPortHasura?: string
  pgPort: string
}

const defaultConfig: Config = {
  adminSecret: process.env.HASURA_ADMIN_SECRET,
  endpoint: process.env.HASURA_ENDPOINT,
  pgHost: process.env.PGHOST,
  pgHostHasura: process.env.PGHOST_HASURA,
  pgPortHasura: process.env.PGPORT_HASURA,
  pgPort: process.env.PGPORT,
};

export default class HasuraClient {
  static DEFAULT_DATABASE = 'default';
  static DEFAULT_SCHEMA = 'public';

  private readonly deps: Dependencies;

  constructor (deps?: Partial<Dependencies>, private readonly config: Config = defaultConfig) {
    this.deps = {
      fetch,
      ...deps,
    };
  }

  async executeSql (sql: string, opts: SqlOptions): Promise<any> {
    const response: Response = await this.deps.fetch(
      `${this.config.endpoint}/v2/query`,
      {
        method: 'POST',
        headers: {
          'X-Hasura-Admin-Secret': this.config.adminSecret,
        },
        body: JSON.stringify({
          type: 'run_sql',
          args: {
            sql,
            read_only: opts.readOnly,
            source: opts.source ?? 'default',
          },
        }),
      }
    );

    const body: string = await response.text();

    if (response.status !== 200) {
      throw new Error(body);
    }

    return JSON.parse(body);
  }

  async executeMetadataRequest (
    type: string,
    args: MetadataRequestArgs,
    version?: number
  ): Promise<any> {
    const response: Response = await this.deps.fetch(
      `${this.config.endpoint}/v1/metadata`,
      {
        method: 'POST',
        headers: {
          'X-Hasura-Admin-Secret': this.config.adminSecret,
        },
        body: JSON.stringify({
          type,
          args,
          ...(version && { version }),
        }),
      }
    );

    const body: string = await response.text();

    if (response.status !== 200) {
      throw new Error(body);
    }

    return JSON.parse(body);
  }

  async executeBulkMetadataRequest (
    metadataRequests: MetadataRequests
  ): Promise<any> {
    return await this.executeMetadataRequest('bulk', metadataRequests);
  }

  async exportMetadata (): Promise<HasuraMetadata> {
    const { metadata } = await this.executeMetadataRequest(
      'export_metadata',
      {},
      2
    );
    return metadata;
  }

  async getDbConnectionParameters (account: string): Promise<HasuraDatabaseConnectionParameters> {
    const metadata: HasuraMetadata = await this.exportMetadata();
    const source = metadata.sources.find((source: HasuraSource) => source.name === account);
    if (source === undefined) {
      throw new Error(`Could not find connection parameters for user ${account} on respective database.`);
    } else if (isDefaultDatabaseUrl(source.configuration.connection_info.database_url)) {
      throw new Error('Default connection parameters are not supported.');
    }
    return source.configuration.connection_info.database_url.connection_parameters;
  }

  async doesSourceExist (source: string): Promise<boolean> {
    const metadata = await this.exportMetadata();
    return metadata.sources.filter(({ name }: { name: string }) => name === source).length > 0;
  }

  async doesSchemaExist (source: string, schemaName: string): Promise<boolean> {
    const { result } = await this.executeSql(
      `SELECT schema_name FROM information_schema.schemata WHERE schema_name = '${schemaName}'`,
      { source, readOnly: true }
    );

    return result.length > 1;
  }

  async createSchema (source: string, schemaName: string): Promise<any> {
    return await this.executeSql(`CREATE schema ${schemaName}`, {
      source,
      readOnly: false,
    });
  }

  async executeSqlOnSchema (source: string, schemaName: string, sqlScript: string): Promise<any> {
    return await this.executeSql(
      `
      set schema '${schemaName}';
      ${sqlScript}
      `,
      { source, readOnly: false }
    );
  }

  async getTableNames (schemaName: string, source: string): Promise<string[]> {
    const tablesInSource = await this.executeMetadataRequest(
      'pg_get_source_tables',
      {
        source,
      }
    );
    return tablesInSource
      .filter(({ schema }: { schema: string }) => schema === schemaName)
      .map(({ name }: { name: string }) => name);
  }

  async getTrackedTablePermissions (
    databaseName: string,
    schemaName: string,
  ): Promise<HasuraTableMetadata[]> {
    const metadata: HasuraMetadata = await this.exportMetadata();
    const hasuraSource = metadata.sources.find((source: HasuraSource) => source.name === databaseName);
    const tablesForSchema = hasuraSource?.tables.filter((tableMetadata: HasuraTableMetadata) => tableMetadata.table.schema === schemaName);

    return tablesForSchema ?? [];
  }

  async trackTables (
    schemaName: string,
    tableNames: string[],
    source: string
  ): Promise<any> {
    return await this.executeBulkMetadataRequest(
      tableNames.map(name => ({
        type: 'pg_track_table',
        args: {
          source,
          table: {
            name,
            schema: schemaName,
          },
        },
      }))
    );
  }

  async untrackTables (
    source: string,
    schema: string,
    tableNames: string[],
    cascade = true
  ): Promise<any> {
    return await this.executeBulkMetadataRequest(
      tableNames.map(name => ({
        type: 'pg_untrack_table',
        args: {
          table: {
            schema,
            name,
          },
          source,
          cascade,
        },
      }))
    );
  }

  async getForeignKeys (schemaName: string, source: string): Promise<any[]> {
    const { result } = await this.executeSql(
      `
      SELECT
        COALESCE(json_agg(row_to_json(info)), '[]'::JSON)
      FROM (
        SELECT
          q.table_schema::text AS table_schema,
          q.table_name::text AS table_name,
          q.constraint_name::text AS constraint_name,
          min(q.ref_table_table_schema::text) AS ref_table_table_schema,
          min(q.ref_table::text) AS ref_table,
          json_object_agg(ac.attname, afc.attname) AS column_mapping,
          min(q.confupdtype::text) AS on_update,
          min(q.confdeltype::text) AS
          on_delete
        FROM (
          SELECT
            ctn.nspname AS table_schema,
            ct.relname AS table_name,
            r.conrelid AS table_id,
            r.conname AS constraint_name,
            cftn.nspname AS ref_table_table_schema,
            cft.relname AS ref_table,
            r.confrelid AS ref_table_id,
            r.confupdtype,
            r.confdeltype,
            unnest(r.conkey) AS column_id,
            unnest(r.confkey) AS ref_column_id
          FROM
            pg_constraint r
            JOIN pg_class ct ON r.conrelid = ct.oid
            JOIN pg_namespace ctn ON ct.relnamespace = ctn.oid
            JOIN pg_class cft ON r.confrelid = cft.oid
            JOIN pg_namespace cftn ON cft.relnamespace = cftn.oid
          WHERE
            r.contype = 'f'::"char"
            AND ((ctn.nspname='${schemaName}'))
            ) q
          JOIN pg_attribute ac ON q.column_id = ac.attnum
            AND q.table_id = ac.attrelid
          JOIN pg_attribute afc ON q.ref_column_id = afc.attnum
            AND q.ref_table_id = afc.attrelid
          GROUP BY
            q.table_schema,
            q.table_name,
            q.constraint_name) AS info;
      `,
      { readOnly: true, source }
    );

    const [, [foreignKeysJsonString]] = result;

    return JSON.parse(foreignKeysJsonString);
  }

  async trackForeignKeyRelationships (
    schemaName: string,
    source: string
  ): Promise<any> {
    const foreignKeys = await this.getForeignKeys(schemaName, source);

    if (foreignKeys.length === 0) {
      return;
    }

    return await this.executeBulkMetadataRequest(
      foreignKeys
        .map((foreignKey) => ([
          {
            type: 'pg_create_array_relationship',
            args: {
              source,
              name: foreignKey.table_name,
              table: {
                name: foreignKey.ref_table,
                schema: schemaName,
              },
              using: {
                foreign_key_constraint_on: {
                  table: {
                    name: foreignKey.table_name,
                    schema: schemaName,
                  },
                  column: Object.keys(foreignKey.column_mapping)[0],
                }
              },
            }
          },
          {
            type: 'pg_create_object_relationship',
            args: {
              source,
              name: pluralize.singular(foreignKey.ref_table),
              table: {
                name: foreignKey.table_name,
                schema: schemaName,
              },
              using: {
                foreign_key_constraint_on: Object.keys(foreignKey.column_mapping)[0],
              },
            }
          },
        ]))
        .flat()
    );
  }

  async addPermissionsToTables (schemaName: string, source: string, tableNames: string[], roleName: string, permissions: string[]): Promise<any> {
    return await this.executeBulkMetadataRequest(
      tableNames
        .map((tableName) => (
          permissions.map((permission) => ({
            type: `pg_create_${permission}_permission`,
            args: {
              source,
              table: {
                name: tableName,
                schema: schemaName,
              },
              role: roleName,
              permission: {
                columns: '*',
                check: {},
                computed_fields: [],
                filter: {},
                ...(permission !== 'select' && { backend_only: true }),
                ...(permission === 'select' && { allow_aggregations: true })
              },
            },
          }))
        ))
        .flat()
    );
  }

  async addDatasource (userName: string, password: string, databaseName: string): Promise<any> {
    return await this.executeMetadataRequest('pg_add_source', {
      name: databaseName,
      configuration: {
        connection_info: {
          database_url: {
            connection_parameters: {
              password,
              database: databaseName,
              username: userName,
              host: this.config.pgHostHasura ?? this.config.pgHost,
              port: Number(this.config.pgPortHasura ?? this.config.pgPort),
            }
          },
        },
      },
    });
  }
}
