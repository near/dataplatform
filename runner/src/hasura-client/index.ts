export { default } from './hasura-client';
export type { HasuraDatabaseConnectionParameters } from './hasura-client';
