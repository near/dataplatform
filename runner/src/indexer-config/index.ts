export { default } from './indexer-config';
export { ProvisioningConfig, LocalIndexerConfig } from './indexer-config';
