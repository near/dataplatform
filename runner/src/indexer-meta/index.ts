export { default } from './indexer-meta';
export { IndexerStatus, METADATA_TABLE_UPSERT, MetadataFields } from './indexer-meta';
