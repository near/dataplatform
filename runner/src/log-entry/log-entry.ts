import { LogType, LogLevel } from '../indexer-logger/indexer-logger';

export default class LogEntry {
  public readonly timestamp: Date;

  constructor (
    public readonly message: string,
    public readonly level: LogLevel,
    public readonly type: LogType,
    public readonly blockHeight?: number
  ) {
    this.timestamp = new Date();
  }

  static createLog (message: string, level: LogLevel, type: LogType, blockHeight?: number): LogEntry {
    return new LogEntry(message, level, type, blockHeight);
  }

  static systemDebug (message: string, blockHeight?: number): LogEntry {
    return LogEntry.createLog(message, LogLevel.DEBUG, LogType.SYSTEM, blockHeight);
  }

  static systemInfo (message: string, blockHeight?: number): LogEntry {
    return LogEntry.createLog(message, LogLevel.INFO, LogType.SYSTEM, blockHeight);
  }

  static systemWarn (message: string, blockHeight?: number): LogEntry {
    return LogEntry.createLog(message, LogLevel.WARN, LogType.SYSTEM, blockHeight);
  }

  static systemError (message: string, blockHeight?: number): LogEntry {
    return LogEntry.createLog(message, LogLevel.ERROR, LogType.SYSTEM, blockHeight);
  }

  static userDebug (message: string, blockHeight?: number): LogEntry {
    return LogEntry.createLog(message, LogLevel.DEBUG, LogType.USER, blockHeight);
  }

  static userInfo (message: string, blockHeight?: number): LogEntry {
    return LogEntry.createLog(message, LogLevel.INFO, LogType.USER, blockHeight);
  }

  static userWarn (message: string, blockHeight?: number): LogEntry {
    return LogEntry.createLog(message, LogLevel.WARN, LogType.USER, blockHeight);
  }

  static userError (message: string, blockHeight?: number): LogEntry {
    return LogEntry.createLog(message, LogLevel.ERROR, LogType.USER, blockHeight);
  }
}
