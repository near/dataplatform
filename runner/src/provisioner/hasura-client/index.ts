export { default } from './hasura-client';
export type { HasuraMetadata, HasuraSource, HasuraConfiguration, HasuraDatabaseConnectionParameters, HasuraTableMetadata, HasuraRolePermission, HasuraPermission } from './hasura-client';
