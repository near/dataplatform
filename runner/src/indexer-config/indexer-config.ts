import crypto from 'crypto';
import { type StartExecutorRequest__Output } from '../generated/runner/StartExecutorRequest';
import { LogLevel } from '../indexer-meta/log-entry';

interface IndexerConfigData {
  redisStreamKey: string
  accountId: string
  functionName: string
  version: number
  code: string
  schema: string
  logLevel: LogLevel
}

export default class IndexerConfig {
  public readonly executorId: string;

  constructor (
    public readonly redisStreamKey: string,
    public readonly accountId: string,
    public readonly functionName: string,
    public readonly version: number,
    public readonly code: string,
    public readonly schema: string,
    public readonly logLevel: LogLevel
  ) {
    const hash = crypto.createHash('sha256');
    hash.update(`${accountId}/${functionName}`);
    this.executorId = hash.digest('hex');
  }

  static fromStartRequest (startExecutorRequest: StartExecutorRequest__Output): IndexerConfig {
    return new IndexerConfig(
      startExecutorRequest.redisStream,
      startExecutorRequest.accountId,
      startExecutorRequest.functionName,
      parseInt(startExecutorRequest.version),
      startExecutorRequest.code,
      startExecutorRequest.schema,
      LogLevel.INFO
    );
  }

  static fromObject (data: IndexerConfigData): IndexerConfig {
    return new IndexerConfig(
      data.redisStreamKey,
      data.accountId,
      data.functionName,
      data.version,
      data.code,
      data.schema,
      data.logLevel
    );
  }

  toObject (): IndexerConfigData {
    return {
      redisStreamKey: this.redisStreamKey,
      accountId: this.accountId,
      functionName: this.functionName,
      version: this.version,
      code: this.code,
      schema: this.schema,
      logLevel: this.logLevel
    };
  }

  private sanitizeNameForDatabase (name: string): string {
    // TODO: Add underscore for accounts with invalid starting character
    return name.replace(/[^a-zA-Z0-9]/g, '_');
  }

  fullName (): string {
    return `${this.accountId}/${this.functionName}`;
  }

  hasuraRoleName (): string {
    return this.sanitizeNameForDatabase(this.accountId);
  }

  hasuraFunctionName (): string {
    return this.sanitizeNameForDatabase(this.functionName);
  }

  userName (): string {
    return this.sanitizeNameForDatabase(this.accountId);
  }

  databaseName (): string {
    return this.sanitizeNameForDatabase(this.accountId);
  }

  schemaName (): string {
    return this.sanitizeNameForDatabase(this.fullName());
  }
}
