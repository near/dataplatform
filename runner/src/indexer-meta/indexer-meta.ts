import format from 'pg-format';
import { wrapError } from '../utility';
import PgClient, { type PostgresConnectionParams } from '../pg-client';
import { trace } from '@opentelemetry/api';
import type LogEntry from './log-entry';
import { LogLevel } from './log-entry';

export enum IndexerStatus {
  PROVISIONING = 'PROVISIONING',
  RUNNING = 'RUNNING',
  FAILING = 'FAILING',
  STOPPED = 'STOPPED',
}

const METADATA_TABLE_UPSERT = 'INSERT INTO %I.__metadata (attribute, value) VALUES %L ON CONFLICT (attribute) DO UPDATE SET value = EXCLUDED.value RETURNING *';
const STATUS_ATTRIBUTE = 'STATUS';
const LAST_PROCESSED_BLOCK_HEIGHT_ATTRIBUTE = 'LAST_PROCESSED_BLOCK_HEIGHT';

export default class IndexerMeta {
  tracer = trace.getTracer('queryapi-runner-indexer-logger');

  private readonly pgClient: PgClient;
  private readonly schemaName: string;
  private readonly logInsertQueryTemplate: string = 'INSERT INTO %I.__logs (block_height, date, timestamp, type, level, message) VALUES %L';
  private readonly loggingLevel: number;

  constructor (
    functionName: string,
    loggingLevel: number,
    databaseConnectionParameters: PostgresConnectionParams,
    pgClientInstance: PgClient | undefined = undefined
  ) {
    const pgClient = pgClientInstance ?? new PgClient(databaseConnectionParameters);

    this.pgClient = pgClient;
    this.schemaName = functionName.replace(/[^a-zA-Z0-9]/g, '_');
    this.loggingLevel = loggingLevel;
  }

  private shouldLog (logLevel: LogLevel): boolean {
    return logLevel >= this.loggingLevel;
  }

  async writeLogs (
    logEntries: LogEntry[],
  ): Promise<void> {
    const entriesArray = logEntries.filter(entry => this.shouldLog(entry.level));
    if (entriesArray.length === 0) return;

    const spanMessage = `write log for ${entriesArray.length === 1 ? 'single entry' : `batch of ${entriesArray.length}`} through postgres `;
    const writeLogSpan = this.tracer.startSpan(spanMessage);
    await wrapError(async () => {
      const values = entriesArray.map(entry => [
        entry.blockHeight,
        entry.timestamp,
        entry.timestamp,
        entry.type,
        LogLevel[entry.level],
        entry.message
      ]);
      const query = format(this.logInsertQueryTemplate, this.schemaName, values);
      await this.pgClient.query(query);
    }, `Failed to insert ${entriesArray.length > 1 ? 'logs' : 'log'} into the ${this.schemaName}.__logs table`)
      .finally(() => {
        writeLogSpan.end();
      });
  }

  async setStatus (status: IndexerStatus): Promise<void> {
    const setStatusSpan = this.tracer.startSpan(`set status of indexer to ${status} through postgres`);
    const values = [[STATUS_ATTRIBUTE, status]];
    const query = format(METADATA_TABLE_UPSERT, this.schemaName, values);

    try {
      await wrapError(async () => await this.pgClient.query(query), `Failed to update status for ${this.schemaName}`);
    } finally {
      setStatusSpan.end();
    }
  }

  async updateBlockheight (blockHeight: number): Promise<void> {
    const setLastProcessedBlockSpan = this.tracer.startSpan(`set last processed block to ${blockHeight} through postgres`);
    const values = [[LAST_PROCESSED_BLOCK_HEIGHT_ATTRIBUTE, blockHeight.toString()]];
    const query = format(METADATA_TABLE_UPSERT, this.schemaName, values);

    try {
      await wrapError(async () => await this.pgClient.query(query), `Failed to update last processed block height for ${this.schemaName}`);
    } finally {
      setLastProcessedBlockSpan.end();
    }
  }
}
