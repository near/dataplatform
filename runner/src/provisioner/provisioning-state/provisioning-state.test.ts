import { ProvisioningConfig } from '../../indexer-config';
import { LogLevel } from '../../indexer-meta/log-entry';
import type HasuraClient from '../hasura-client';
import { type HasuraTableMetadata, type HasuraConfiguration, type HasuraDatabaseConnectionParameters, type HasuraSource, type HasuraMetadata } from '../hasura-client';
import ProvisioningState from './provisioning-state';

describe('ProvisioiningState', () => {
  const provisioningConfig = new ProvisioningConfig(
    'account-id',
    'function-name',
    'schema',
    LogLevel.INFO,
  );

  it('can create state whether source exists or not', async () => {
    const metadataWithoutUser = generateHasuraMetadata(['some_schema'], ['tableA', 'tableB'], 'someAccount', 'someDb');
    const mockExportMetadata = jest.fn().mockResolvedValue(metadataWithoutUser);
    const mockGetTableNames = jest.fn().mockResolvedValue([]);
    const mockHasuraClient = {
      exportMetadata: mockExportMetadata,
      getTableNames: mockGetTableNames,
    } as unknown as HasuraClient;

    const provisioningState = await ProvisioningState.loadProvisioningState(mockHasuraClient, provisioningConfig);
    expect(provisioningState.doesSourceExist()).toBe(false);
    expect(provisioningState.doesSchemaExist()).toBe(false);
    expect(provisioningState.getCreatedTables()).toEqual([]);
  });

  it('state works with existing source', async () => {
    const metadataWithUser = generateHasuraMetadata([provisioningConfig.schemaName(), 'some_schema'], ['tableA', 'tableB'], provisioningConfig.hasuraRoleName(), provisioningConfig.databaseName());
    metadataWithUser.sources.push(generateSourceWithTables(['anotherSchema'], ['anotherTable'], 'anotherRole', 'anotherDb'));
    const mockExportMetadata = jest.fn().mockResolvedValue(metadataWithUser);
    const mockGetTableNames = jest.fn().mockResolvedValue(['tableA']);
    const mockHasuraClient = {
      exportMetadata: mockExportMetadata,
      getTableNames: mockGetTableNames,
    } as unknown as HasuraClient;

    const provisioningState = await ProvisioningState.loadProvisioningState(mockHasuraClient, provisioningConfig);
    expect(provisioningState.doesSourceExist()).toBe(true);
    expect(provisioningState.doesSchemaExist()).toBe(true);
    expect(provisioningState.getCreatedTables()).toEqual(['tableA']);
  });

  it('correctly fetch metadata for existing source and schema', async () => {
    const metadataWithUser = generateHasuraMetadata([provisioningConfig.schemaName(), 'some_schema'], ['tableA', 'tableB'], provisioningConfig.hasuraRoleName(), provisioningConfig.databaseName());
    metadataWithUser.sources.push(generateSourceWithTables(['anotherSchema'], ['anotherTable'], 'anotherRole', 'anotherDb'));
    const mockExportMetadata = jest.fn().mockResolvedValue(metadataWithUser);
    const mockGetTableNames = jest.fn().mockResolvedValue(['tableA']);
    const mockHasuraClient = {
      exportMetadata: mockExportMetadata,
      getTableNames: mockGetTableNames,
    } as unknown as HasuraClient;

    const provisioningState = await ProvisioningState.loadProvisioningState(mockHasuraClient, provisioningConfig);
    expect(provisioningState.getSourceMetadata().name).toBe(provisioningConfig.hasuraRoleName());
    expect(provisioningState.getMetadataForTables().length).toBe(2);
    expect(provisioningState.getMetadataForTables()).toMatchSnapshot();
    expect(provisioningState.getTrackedTables()).toEqual(['tableA', 'tableB']);
    expect(provisioningState.getTablesWithPermissions()).toEqual(['tableA', 'tableB']);
  });

  it('handles table with missing permissions', async () => {
    const metadataWithUser = generateHasuraMetadata([provisioningConfig.schemaName(), 'some_schema'], ['tableA', 'tableB'], provisioningConfig.hasuraRoleName(), provisioningConfig.databaseName());
    const role = provisioningConfig.hasuraRoleName();
    const tableMissingPermissions = {
      table: {
        name: 'tableC',
        schema: provisioningConfig.schemaName(),
      },
      insert_permissions: [{ role, permission: {} }],
      update_permissions: [{ role, permission: {} }],
      delete_permissions: [{ role, permission: {} }],
    };
    const tableWithIncorrectlyNamedPermission = {
      table: {
        name: 'tableD',
        schema: provisioningConfig.schemaName(),
      },
      select_permissions: [{ role, permission: {} }],
      insert_permission: [{ role, permission: {} }],
      update_permissions: [{ role, permission: {} }],
      delete_permissions: [{ role, permission: {} }],
    };
    metadataWithUser.sources[1].tables.push(tableMissingPermissions); // First source is a default source
    metadataWithUser.sources[1].tables.push(tableWithIncorrectlyNamedPermission);

    const mockExportMetadata = jest.fn().mockResolvedValue(metadataWithUser);
    const mockGetTableNames = jest.fn().mockResolvedValue(['tableA']);
    const mockHasuraClient = {
      exportMetadata: mockExportMetadata,
      getTableNames: mockGetTableNames,
    } as unknown as HasuraClient;

    const provisioningState = await ProvisioningState.loadProvisioningState(mockHasuraClient, provisioningConfig);
    expect(provisioningState.getSourceMetadata().name).toBe(provisioningConfig.hasuraRoleName());
    expect(provisioningState.getMetadataForTables().length).toBe(4);
    expect(provisioningState.getMetadataForTables()).toMatchSnapshot();
    expect(provisioningState.getTrackedTables()).toEqual(['tableA', 'tableB', 'tableC', 'tableD']);
    expect(provisioningState.getTablesWithPermissions()).toEqual(['tableA', 'tableB']);
  });

  it('throws error when multiple sources with same name exist', async () => {
    const metadataWithUser = generateHasuraMetadata([provisioningConfig.schemaName(), 'some_schema'], ['tableA', 'tableB'], provisioningConfig.hasuraRoleName(), provisioningConfig.databaseName());
    metadataWithUser.sources.push(generateSourceWithTables(['anotherSchema'], ['anotherTable'], provisioningConfig.hasuraRoleName(), provisioningConfig.databaseName()));
    const mockExportMetadata = jest.fn().mockResolvedValue(metadataWithUser);
    const mockGetTableNames = jest.fn().mockResolvedValue(['tableA']);
    const mockHasuraClient = {
      exportMetadata: mockExportMetadata,
      getTableNames: mockGetTableNames,
    } as unknown as HasuraClient;

    const provisioningState = await ProvisioningState.loadProvisioningState(mockHasuraClient, provisioningConfig);
    expect(() => provisioningState.getSourceMetadata()).toThrow('Expected exactly one source');
    expect(() => provisioningState.getMetadataForTables()).toThrow('Expected exactly one source');
    expect(() => provisioningState.getTrackedTables()).toThrow('Expected exactly one source');
  });
});

function generateHasuraMetadata (schemaNames: string[], tableNames: string[], role: string, db: string): HasuraMetadata {
  const sources: HasuraSource[] = [];
  // Insert default source which has different format than the rest
  sources.push({
    name: 'default',
    kind: 'postgres',
    tables: [],
    configuration: {
      connection_info: {
        database_url: { from_env: 'HASURA_GRAPHQL_DATABASE_URL' },
      }
    }
  });

  sources.push(generateSourceWithTables(schemaNames, tableNames, role, db));

  return {
    version: 3,
    sources
  };
}

function generateSourceWithTables (schemaNames: string[], tableNames: string[], role: string, db: string): HasuraSource {
  const tables: HasuraTableMetadata[] = [];
  schemaNames.forEach((schemaName) => {
    tableNames.forEach((tableName) => {
      tables.push(generateTableConfig(schemaName, tableName, role));
    });
  });

  return {
    name: db,
    kind: 'postgres',
    tables,
    configuration: generateHasuraConfiguration(role, 'password'),
  };
}

function generateTableConfig (schemaName: string, tableName: string, role: string): HasuraTableMetadata {
  return {
    table: {
      name: tableName,
      schema: schemaName,
    },
    insert_permissions: [{ role, permission: {} }],
    select_permissions: [{ role, permission: {} }],
    update_permissions: [{ role, permission: {} }],
    delete_permissions: [{ role, permission: {} }],
  };
}

function generateHasuraConfiguration (user: string, password: string): HasuraConfiguration {
  return {
    connection_info: {
      database_url: { connection_parameters: generateConnectionParameter(user, password) },
      isolation_level: 'read-committed',
      use_prepared_statements: false
    }
  };
}

function generateConnectionParameter (user: string, password: string): HasuraDatabaseConnectionParameters {
  return {
    database: user,
    host: 'postgres',
    password,
    port: 5432,
    username: user
  };
}
