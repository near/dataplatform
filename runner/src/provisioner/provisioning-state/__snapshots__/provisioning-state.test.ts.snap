// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProvisioiningState correctly fetch metadata for existing source and schema 1`] = `
[
  {
    "delete_permissions": [
      {
        "permission": {},
        "role": "account_id",
      },
    ],
    "insert_permissions": [
      {
        "permission": {},
        "role": "account_id",
      },
    ],
    "select_permissions": [
      {
        "permission": {},
        "role": "account_id",
      },
    ],
    "table": {
      "name": "tableA",
      "schema": "account_id_function_name",
    },
    "update_permissions": [
      {
        "permission": {},
        "role": "account_id",
      },
    ],
  },
  {
    "delete_permissions": [
      {
        "permission": {},
        "role": "account_id",
      },
    ],
    "insert_permissions": [
      {
        "permission": {},
        "role": "account_id",
      },
    ],
    "select_permissions": [
      {
        "permission": {},
        "role": "account_id",
      },
    ],
    "table": {
      "name": "tableB",
      "schema": "account_id_function_name",
    },
    "update_permissions": [
      {
        "permission": {},
        "role": "account_id",
      },
    ],
  },
]
`;

exports[`ProvisioiningState handles table with missing permissions 1`] = `
[
  {
    "delete_permissions": [
      {
        "permission": {},
        "role": "account_id",
      },
    ],
    "insert_permissions": [
      {
        "permission": {},
        "role": "account_id",
      },
    ],
    "select_permissions": [
      {
        "permission": {},
        "role": "account_id",
      },
    ],
    "table": {
      "name": "tableA",
      "schema": "account_id_function_name",
    },
    "update_permissions": [
      {
        "permission": {},
        "role": "account_id",
      },
    ],
  },
  {
    "delete_permissions": [
      {
        "permission": {},
        "role": "account_id",
      },
    ],
    "insert_permissions": [
      {
        "permission": {},
        "role": "account_id",
      },
    ],
    "select_permissions": [
      {
        "permission": {},
        "role": "account_id",
      },
    ],
    "table": {
      "name": "tableB",
      "schema": "account_id_function_name",
    },
    "update_permissions": [
      {
        "permission": {},
        "role": "account_id",
      },
    ],
  },
  {
    "delete_permissions": [
      {
        "permission": {},
        "role": "account_id",
      },
    ],
    "insert_permissions": [
      {
        "permission": {},
        "role": "account_id",
      },
    ],
    "table": {
      "name": "tableC",
      "schema": "account_id_function_name",
    },
    "update_permissions": [
      {
        "permission": {},
        "role": "account_id",
      },
    ],
  },
  {
    "delete_permissions": [
      {
        "permission": {},
        "role": "account_id",
      },
    ],
    "insert_permission": [
      {
        "permission": {},
        "role": "account_id",
      },
    ],
    "select_permissions": [
      {
        "permission": {},
        "role": "account_id",
      },
    ],
    "table": {
      "name": "tableD",
      "schema": "account_id_function_name",
    },
    "update_permissions": [
      {
        "permission": {},
        "role": "account_id",
      },
    ],
  },
]
`;
